### PR TITLE
gaurav's-progressive-web-app 

### PR DESCRIPTION
**Pull request by gauravThakur-sgl**
### Feature Description
It Progressive Web App (PWA) compatibility to our application, enhancing user experience and providing offline functionality.

###  Changes Made
- Added PWA support with `serviceWorker.js` implementation
- Created install prompt mechanism
- Updated `manifest.json`for PWA requirements
- Implemented install button for app installation prompt

### Screenshot
Screenshot from 2024-11-27 16-55-37

### CODE REVIEW

**Comment by gauravThakur-sgl** [comment]
@shubhamships Sir
### Changes Made
- used local state to manipulate the states
- created and state with empty array `  const [punchData, setPunchData] = useState<{ id: number; time: string; type: string }[]>([]);`
- during `if (punchData.length === 0 || punchData[punchData.length - 1].type === "OUT") {
        setPunchData([
          ...punchData,
          {
            id: punchData.length,
            time: new Date().toISOString(),
            type: "IN",
          },
        ]);
        setIsDisabled(true);
      }` 
        to the empty array

 - During `punchOutData`  pushed the 
  `if (punchData.length === 0 || punchData[punchData.length - 1].type === "IN") {
        setPunchData([
          ...punchData,
          {
            id: punchData.length,
            time: new Date().toISOString(),
            type: "OUT",
          },
        ]);
      }`
  - Rendered the `punchInData` based on the type of Data
      
      #### And all the logic remained same
      
      ##### punchData Details
 
`{id: 0, time: '2024-11-28T18:44:11.813Z', type: 'IN'}
{id: 1, time: '2024-11-28T18:44:12.796Z', type: 'OUT'} 
{id: 2, time: '2024-11-28T18:55:50.847Z', type: 'IN'}
{id: 3, time: '2024-11-28T18:55:54.162Z', type: 'OUT'}
length: 4`

**Comment by gauravThakur-sgl** [comment]
@shubhamships Sir
### Changes Made
- Removed the previous logic for checking `totalTime < 9 `
- Calculated total time `totalTime = (punchOutTime - punchInTime) / 1000 ` in ms
- checked the the total time using condition 
`Total time elapsed: ${
              netTime < 60
                ? netTime.toFixed(2) + " second"
                : netTime < 3600
                ? (netTime / 60).toFixed(2) + " minutes"
                : (netTime / 3600).toFixed(2) + " hours"
            } `

**Review by shubhamships — APPROVED** [review]

**Review by ChinmaySingh-ShipGlobal — APPROVED** [review]

File: gaurav/public/manifest.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+  "short_name": "Time Management",
+  "name": "Install TimeManagement App",
+  "icons": [
+    {
+      "src": "/clock.svg",
+      "sizes": "192x192",
+      "type": "image/svg+xml"
```
[the remark this answers is not in the document]
Comment (gauravThakur-sgl) [reply gaurav/public/manifest.json:8]: yes

File: gaurav/src/components/MarkTime.tsx [hunk 2]
```diff
@@ -1,90 +1,127 @@
...
-        <h1 className="text-3xl font-semibold text-primary m-4">TimeLog</h1>
+        <h1 className="text-3xl font-semibold text-primary m-4">Time Log</h1>
         <Card className="relative flex flex-col items-center h-full min-w-80 max-w-96 py-8 px-8 gap-2 bg-gray-50 shadow-sm overflow-auto">
+          <div className="text-sm font-semibold text-primary contrast-200 sticky">
+            {`Total time elapsed: ${
+              netTime < 60
+                ? netTime.toFixed(2) + " second"
+                : netTime < 3600
+                ? (netTime / 60).toFixed(2) + " minutes"
+                : (netTime / 3600).toFixed(2) + " hours"
```
[the remark this answers is not in the document]
Comment (gauravThakur-sgl) [reply gaurav/src/components/MarkTime.tsx]: fixed

**Pull request merged**